p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,790 @@
+		http.Error(w, "Only application/json is supported", http.StatusNotAcceptable)
+		return
+	}
+
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET requests are supported", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Return a JSON object with the well-known protocols
+	h.wellknownMapMu.Lock()
+	mapping, err := json.Marshal(h.wellKnownMapping)
+	h.wellknownMapMu.Unlock()
+	if err != nil {
+		http.Error(w, "Marshal error", http.StatusInternalServerError)
+		return
+	}
+	w.Header().Add("Content-Type", "application/json")
+	w.Header().Add("Content-Length", strconv.Itoa(len(mapping)))
+	w.Write(mapping)
+}
+
+func (h *WellKnownHandler) AddProtocolMeta(p protocol.ID, protocolMeta ProtocolMeta) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping == nil {
+		h.wellKnownMapping = make(map[protocol.ID]ProtocolMeta)
+	}
+	h.wellKnownMapping[p] = protocolMeta
+	h.wellknownMapMu.Unlock()
+}
+
+func (h *WellKnownHandler) RemoveProtocolMeta(p protocol.ID) {
+	h.wellknownMapMu.Lock()
+	if h.wellKnownMapping != nil {
+		delete(h.wellKnownMapping, p)
+	}
+	h.wellknownMapMu.Unlock()
+}
+
+// Host is a libp2p host for request/responses with HTTP semantics. This is
+// in contrast to a stream-oriented host like the core host.Host interface. Its
+// zero-value (&Host{}) is usable. Do not copy by value.
+// See examples for usage.
+//
+//	Warning, this is experimental. The API will likely change.
+type Host struct {
+	// StreamHost is a stream based libp2p host used to do HTTP over libp2p streams. May be nil
+	StreamHost host.Host
+	// ListenAddrs are the requested addresses to listen on. Multiaddrs must be
+	// valid HTTP(s) multiaddr. Only multiaddrs for an HTTP transport are
+	// supported (must end with /http or /https).
+	ListenAddrs []ma.Multiaddr
+	// TLSConfig is the TLS config for the server to use
+	TLSConfig *tls.Config
+	// InsecureAllowHTTP indicates if the server is allowed to serve unencrypted
+	// HTTP requests over TCP.
+	InsecureAllowHTTP bool
+	// ServeMux is the http.ServeMux used by the server to serve requests. If nil,
+	// new serve mux will be allocated. Users may manually add handlers to this
+	// mux instead of using `SetHTTPHandler`, but if they do, they should also
+	// update the WellKnownHandler's protocol mapping.
+	ServeMux           *http.ServeMux
+	initializeServeMux sync.Once
+
+	// DefaultClientRoundTripper is the default http.RoundTripper for clients to
+	// use when making requests over an HTTP transport. This must be an
+	// `*http.Transport` type so that the transport can be cloned and the
+	// `TLSClientConfig` field can be configured. If unset, it will create a new
+	// `http.Transport` on first use.
+	DefaultClientRoundTripper *http.Transport
+
+	// WellKnownHandler is the http handler for the `.well-known/libp2p`
+	// resource. It is responsible for sharing this node's protocol metadata
+	// with other nodes.
```

Comment:
Mention that users only need to care about it if they set their own `ServeMux`, and that it's handled automatically otherwise (if I understand correctly).